Add TLS support for Inference Loggers

ruivieira opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kserve/kserve/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

This PR adds support for TLS in Inference Loggers.
It adds two new optional fields to the global ConfigMap: `tlsBundle` and `tlsCertName`.

If a ConfigMap matching `tlsBundle` is present in the Inference Service's namespace, it will be mounted as a volume on the agent. 
* If the logger's endpoint protocol is of type `https`, the logger will use the CA cert specified with `tlsCertName` to configure the transport.
  * If TLS is configured in ConfigMap, but the bundle is not available or the CA cert is not found, the logger will continue to behave as before this PR (no TLS transport configured)
* If the logger's endpoint protocol is of type `http`, the inference logger will behave as previously

_why we need it_

The current implementation does not configure HTTP transport for TLS.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes https://github.com/kserve/kserve/issues/3837

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] Exploratory testing on with development deployments
- [x] Agent injection test

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
TLS support for Inference Logger by specifying certificate bundle in the global configuration
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

ruivieira committed:
Initial TLS bundle support

Signed-off-by: Rui Vieira <ruidevieira@googlemail.com>

yuzisun commented on cmd/agent/main.go:
```diff
@@ -74,6 +74,7 @@ var (
 	readinessProbeTimeout = flag.Duration("probe-period", -1, "run readiness probe with given timeout") //nolint: unused
 	// This creates an abstract socket instead of an actual file.
 	unixSocketPath = "@/kserve/agent.sock"
+	tlsCertName    = flag.String("log-tls-cert", "service-ca.crt", "The TLS certificate name")
```
How are we mounting the ca certificate to the container?

yuzisun replied:
Also this is a file not a name

yuzisun commented on cmd/agent/main.go:
```diff
@@ -286,6 +288,7 @@ func startLogger(workers int, logger *zap.SugaredLogger) *loggerArgs {
 		endpoint:         *endpoint,
 		namespace:        *namespace,
 		component:        *component,
+		tlsCertName:      *tlsCertName,
```
to be precise this is CA certificate

ruivieira replied:
Renamed to `caCertName`

yuzisun commented on pkg/webhook/admission/pod/agent_injector.go:
```diff
@@ -56,6 +60,8 @@ type LoggerConfig struct {
 	MemoryRequest string `json:"memoryRequest"`
 	MemoryLimit   string `json:"memoryLimit"`
 	DefaultUrl    string `json:"defaultUrl"`
+	TlsBundle     string `json:"tlsBundle"`
```
Suggest renaming to CABundle

ruivieira replied:
Renamed to `CaBundle`

yuzisun commented on pkg/webhook/admission/pod/agent_injector.go:
```diff
@@ -277,6 +288,30 @@ func (ag *AgentInjector) InjectAgent(pod *v1.Pod) error {
+		configMapVolume := v1.VolumeSource{
+			ConfigMap: &v1.ConfigMapVolumeSource{
+				LocalObjectReference: v1.LocalObjectReference{
+					Name: ag.loggerConfig.TlsBundle,
```
Is the CA bundle configmap available in each namespace?

ruivieira replied:
Hi @yuzisun, addressing this question and the one [above](https://github.com/kserve/kserve/pull/3863#discussion_r1720545006).

The CA bundle's name is specified in the global ConfigMap.
When the agent is created, a ConfigMap from the same namespace as the InferenceService and with that matching name is mounted as a volume and used to configure the transport.
Since the previous behaviour (before this PR) was that `https` could be specified as a logger endpoint, although with no TLS configuration, the following is expected in order keep this backwards compatible with the previous behaviour:
- If the ConfigMap is _not_ present in the InferenceService's namespace and the logger endpoint is `https`, continue using that endpoint but with no TLS configured, i.e. the ConfigMap is optional and it won't cause the agent to fail if missing

spolti commented on pkg/logger/worker.go:
```diff
@@ -85,6 +90,26 @@ func (w *Worker) sendCloudEvent(logReq LogRequest) error {
 
+	if logReq.Url.Scheme == "https" {
+		caCertFile, err := os.ReadFile(filepath.Join(CaCertPath, logReq.TlsCertName))
+		// Do not fail if certificates not found, for backwards compatibility
```
maybe log a warning might be good, missing the cert might be not intentional.

ruivieira replied:
Added log warning for the case when logger endpoint is https but CA cert file not found

ruivieira commented:
@yuzisun @spolti Is there anything still needing to be addressed from the review's comments? Thank you!

terrytangyuan commented:
Can you fix the DCO check?

yuzisun commented on pkg/webhook/admission/pod/agent_injector.go:
```diff
@@ -39,6 +40,9 @@ const (
 	LoggerArgumentEndpoint         = "--endpoint"
 	LoggerArgumentComponent        = "--component"
+	LoggerCaBundleVolume           = "agent-ca-bundle"
+	LoggerCaCertMountPath          = "/etc/tls/logger"
```
move the path to a constant, it is used in multiple places.

yuzisun commented on pkg/logger/worker.go:
```diff
@@ -85,6 +90,29 @@ func (w *Worker) sendCloudEvent(logReq LogRequest) error {
+				TLSClientConfig: &tls.Config{
+					RootCAs:            clientCertPool,
+					MinVersion:         tls.VersionTLS12,
+					InsecureSkipVerify: false,
```
Should we also consider proving an option to skip TLS verification?

yuzisun commented on pkg/webhook/admission/pod/agent_injector_test.go:
```diff
@@ -55,6 +56,12 @@ var (
 		Image:      "gcr.io/kfserving/agent:latest",
 		DefaultUrl: "http://httpbin.org/",
 	}
+	loggerConfigTls = &LoggerConfig{
```
```suggestion
	loggerTLSConfig = &LoggerConfig{
```

Jooho commented on pkg/logger/types.go:
```diff
@@ -32,4 +32,5 @@ type LogRequest struct {
 	Component        string
 	Endpoint         string
 	Metadata         map[string][]string
+	TlsCertName      string
```
Is it a server certificate? so I think certName is ok.

```suggestion
	certName      string
```

Jooho commented on pkg/logger/handler.go:
```diff
@@ -43,10 +43,11 @@ type LoggerHandler struct {
 	endpoint         string
 	next             http.Handler
 	metadataHeaders  []string
+	tlsCertName      string
```
this is the same 
```suggestion
	certName      string
```

Jooho commented on pkg/logger/handler.go:
```diff
@@ -43,10 +43,11 @@ type LoggerHandler struct {
 
 func New(logUrl *url.URL, sourceUri *url.URL, logMode v1beta1.LoggerType,
-	inferenceService string, namespace string, endpoint string, component string, next http.Handler, metadataHeaders []string) http.Handler {
+	inferenceService string, namespace string, endpoint string, component string, next http.Handler, metadataHeaders []string, tlsCertName string) http.Handler {
```
```suggestion
	inferenceService string, namespace string, endpoint string, component string, next http.Handler, metadataHeaders []string, certName string) http.Handler {
```

Jooho commented on pkg/logger/handler.go:
```diff
@@ -59,6 +60,7 @@ func New(logUrl *url.URL, sourceUri *url.URL, logMode v1beta1.LoggerType,
 		endpoint:         endpoint,
 		next:             next,
 		metadataHeaders:  metadataHeaders,
+		tlsCertName:      tlsCertName,
```
same 
```suggestion
		certName:      certName,
```

Jooho commented on pkg/logger/handler.go:
```diff
@@ -113,6 +115,7 @@ func (eh *LoggerHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			Endpoint:         eh.endpoint,
 			Component:        eh.component,
 			Metadata:         metadata,
+			TlsCertName:      eh.tlsCertName,
```
```suggestion
			certName:      eh.certName,
```

Jooho commented on pkg/logger/handler.go:
```diff
@@ -141,6 +144,7 @@ func (eh *LoggerHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 				Namespace:        eh.namespace,
 				Endpoint:         eh.endpoint,
 				Component:        eh.component,
+				TlsCertName:      eh.tlsCertName,
```
```suggestion
				certName:      eh.certName,
```

ruivieira committed:
Move CA bundle volume and mount path to constants

Signed-off-by: Rui Vieira <ruidevieira@googlemail.com>

ruivieira committed:
Rename loggerConfigTls to loggerTLSConfig

Signed-off-by: Rui Vieira <ruidevieira@googlemail.com>

ruivieira committed:
Rename TlsCertName to CertName

Signed-off-by: Rui Vieira <ruidevieira@googlemail.com>

terrytangyuan reviewed (commented):
PTAL @yuzisun

ruivieira commented:
@terrytangyuan This is don't have the support for skipping TLS verification, I hope to push that today and comment here when done.
@yuzisun @terrytangyuan Another question, at the moment this PR does include a default value for the cert fields in the ConfigMap. Is this something that should be added as part of this PR, or left without a default CM?

ruivieira committed:
Add Logger option skip TLS verification

Also:
- Fixed incorrect cert name argument name (`--log-tls-cert` is now `--logger-ca-cert-file`)

Signed-off-by: Rui Vieira <ruidevieira@googlemail.com>

ruivieira commented:
@terrytangyuan @yuzisun The changes are ready for review now.

terrytangyuan commented:
We can probably add default later. WDYT? @yuzisun

ruivieira committed:
Correct case

Change CABundle and CACertfile to caBundle and caCertFile.

Signed-off-by: Rui Vieira <ruidevieira@googlemail.com>

yuzisun reviewed (dismissed):
/lgtm
/approve

yuzisun commented:
@ruivieira Can you help update the logger doc for this on the kserve website repo?

ruivieira committed:
Fix linting errors

- Restore newline at the end of charts/kserve-resources/README.md
- Remove import of github.com/kserve/kserve/pkg/constants from `worker.go` and replace with local constant for the CA mount path
- `InsecureSkipVerify: logReq.TlsSkipVerify` was triggering gosec's G402 with "potential 'true' for `logReq.TlsSkipVerify`". Since this value is allowed to be true, this specific line was excluded from the checks and an explanatory comment added
- Remove import of `k8s.io/utils/ptr` and replace with a pointer `&optionalVolume`

Signed-off-by: Rui Vieira <ruidevieira@googlemail.com>

ruivieira commented:
@yuzisun @terrytangyuan Had to add a new commit to fix the linting errors:

- Restore newline at the end of charts/kserve-resources/README.md
- Remove import of github.com/kserve/kserve/pkg/constants from `worker.go` and replace with local constant for the CA mount path
- `InsecureSkipVerify: logReq.TlsSkipVerify` was triggering gosec's G402 with "potential 'true' for `logReq.TlsSkipVerify`". Since this value is allowed to be true, this specific line was excluded from the checks and an explanatory comment added
- Remove import of `k8s.io/utils/ptr` and replace with a pointer `&optionalVolume`

I'm unsure about the gosec error in particular, since if InsecureSkipVerify is to be configurable it can by definition be `true` and will always trigger this warning.

ruivieira commented:
> @ruivieira Can you help update the logger doc for this on the kserve website repo?

Of course :+1:

ruivieira committed:
Fix import sort order on `worker.go`

Signed-off-by: Rui Vieira <ruidevieira@googlemail.com>

terrytangyuan commented:
/lgtm
